src/tree/updater_quantile_hist.cc

```diff
@@ -732,198 +770,237 @@ void QuantileHistMaker::Builder::EvaluateSplit(const std::vector<ExpandEntry>& n
+    const uint32_t offset, const int32_t split_cond,
+    const size_t istart, const size_t iend, size_t *p_left,
+    size_t *p_right) {
+  size_t ileft = 0;
+  size_t iright = 0;
 
-  const auto& rowset = row_set_collection_[nid];
+  const uint32_t max_val = std::numeric_limits<uint32_t>::max();
 
-  Column column = column_matrix.GetColumn(fid);
-  if (column.GetType() == xgboost::common::kDenseColumn) {
-    ApplySplitDenseData(rowset, gmat, &row_split_tloc_, column, split_cond,
-                        default_left);
-  } else {
-    ApplySplitSparseData(rowset, gmat, &row_split_tloc_, column, lower_bound,
-                         upper_bound, split_cond, default_left);
+  for (size_t i = istart; i < iend; i++) {
+    if (idx[rid[i]] == max_val) {
+      if (default_left) {
+        p_left[ileft++] = rid[i];
+      } else {
+        p_right[iright++] = rid[i];
+      }
+    } else {
+      if (static_cast<int32_t>(idx[rid[i]] + offset) <= split_cond) {
+        p_left[ileft++] = rid[i];
+      } else {
+        p_right[iright++] = rid[i];
+      }
+    }
   }
 
-  row_set_collection_.AddSplit(
-      nid, row_split_tloc_, (*p_tree)[nid].LeftChild(), (*p_tree)[nid].RightChild());
-  builder_monitor_.Stop("ApplySplit");
+  return { ileft, iright };
 }
 
-void QuantileHistMaker::Builder::ApplySplitDenseData(
-    const RowSetCollection::Elem rowset,
-    const GHistIndexMatrix& gmat,
-    std::vector<RowSetCollection::Split>* p_row_split_tloc,
-    const Column& column,
-    bst_int split_cond,
-    bool default_left) {
-  std::vector<RowSetCollection::Split>& row_split_tloc = *p_row_split_tloc;
-  constexpr int kUnroll = 8;  // loop unrolling factor
-  const size_t nrows = rowset.end - rowset.begin;
-  const size_t rest = nrows % kUnroll;
-
-#pragma omp parallel for num_threads(nthread_) schedule(static)
-  for (bst_omp_uint i = 0; i < nrows - rest; i += kUnroll) {
-    const bst_uint tid = omp_get_thread_num();
-    auto& left = row_split_tloc[tid].left;
-    auto& right = row_split_tloc[tid].right;
-    size_t rid[kUnroll];
-    uint32_t rbin[kUnroll];
-    for (int k = 0; k < kUnroll; ++k) {
-      rid[k] = rowset.begin[i + k];
-    }
-    for (int k = 0; k < kUnroll; ++k) {
-      rbin[k] = column.GetFeatureBinIdx(rid[k]);
-    }
-    for (int k = 0; k < kUnroll; ++k) {                      // NOLINT
-      if (rbin[k] == std::numeric_limits<uint32_t>::max()) {  // missing value
-        if (default_left) {
-          left.push_back(rid[k]);
-        } else {
-          right.push_back(rid[k]);
+template<bool default_left>
+inline std::pair<size_t, size_t> PartitionSparseKernel(const size_t* rid,
+    const uint32_t* idx, const uint32_t offset, const int32_t split_cond,
+    const size_t istart, const size_t iend, size_t* p_left, size_t* p_right,
+    bst_uint lower_bound, bst_uint upper_bound, const Column& column) {
+
+  size_t ileft = 0;
+  size_t iright = 0;
+
+  if (istart < iend) {  // ensure that [istart, iend) is nonempty range
```

Comment:
More hard to read code, anyway to make this function more clear while keeping performance?

Reply:
The same as above